test/integration/etcd_sharding_test.go

```diff
@@ -0,0 +1,148 @@
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Create an etcd client to talk to the openshift etcd
+	osEtcdClient, err := originetcd.MakeNewEtcdClient(openShiftMasterConfig.EtcdClientInfo)
+	if err != nil {
+		t.Fatalf("failed to connect to openshift etcd: %v", err)
+	}
+
+	kubeClient, err := testutil.GetClusterAdminKubeClient(adminConfigFile)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Create a project
+	project := &projectapi.Project{
+		ObjectMeta: kapi.ObjectMeta{
+			Name: "shard-project",
+		},
+	}
+	projectResult, err := osClient.Projects().Create(project)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if projectResult.Status.Phase != kapi.NamespaceActive {
+		t.Fatalf("project status is: %s", projectResult.Status.Phase)
+	}
+
+	// Login as a user
+	_, _, _, err = testutil.GetClientForUser(*adminConfig, "shard")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	_, _, _, err = testutil.GetClientForServiceAccount(kubeClient, *adminConfig, "default", "sa")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// Create a template
+	template := &templateapi.Template{
+		Parameters: []templateapi.Parameter{
+			{
+				Name:  "NAME",
+				Value: "shard-template",
+			},
+		},
+	}
+
+	templateObjects := []runtime.Object{
+		&v1.Service{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      "shard-tester",
+				Namespace: "default",
+			},
+			Spec: v1.ServiceSpec{
+				ClusterIP:       "1.2.3.4",
+				SessionAffinity: "sharding",
+			},
+		},
+	}
+	templateapi.AddObjectsToTemplate(template, templateObjects, v1.SchemeGroupVersion)
+
+	obj, err := osClient.TemplateConfigs("default").Create(template)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if len(obj.Objects) != 1 {
+		t.Fatalf("unexpected object: %#v", obj)
+	}
+
+	// Verify data is placed in the correct etcd instances
+	kubeStoragePrefix := openShiftMasterConfig.EtcdStorageConfig.KubernetesStoragePrefix
+	osStoragePrefix := openShiftMasterConfig.EtcdStorageConfig.OpenShiftStoragePrefix
+	kubeKeys := client.NewKeysAPI(kubeEtcdClient)
+	osKeys := client.NewKeysAPI(osEtcdClient)
+	_, err = kubeKeys.Get(context.TODO(), fmt.Sprintf("/%s", kubeStoragePrefix), nil)
+	if err != nil {
+		t.Fatalf("failed to find kuberenetes prefix '%s' in kubernetes etcd instance: %v", kubeStoragePrefix, err)
+	}
+	_, err = kubeKeys.Get(context.TODO(), fmt.Sprintf("/%s", osStoragePrefix), nil)
+	if err == nil {
+		t.Fatalf("found openshift prefix '%s' in kubernetes etcd instance: %v", osStoragePrefix, err)
+	}
+	_, err = osKeys.Get(context.TODO(), fmt.Sprintf("/%s", osStoragePrefix), nil)
+	if err != nil {
+		t.Fatalf("failed to find openshift prefix '%s' in openshift etcd instance: %v", osStoragePrefix, err)
+	}
+	_, err = osKeys.Get(context.TODO(), fmt.Sprintf("/%s", kubeStoragePrefix), nil)
+	if err == nil {
+		t.Fatalf("found kuberenetes prefix '%s' in openshift etcd instance: %v", kubeStoragePrefix, err)
+	}
+}
```

Comment:
still needs the project deletion at the end, and waiting to ensure the project actually gets cleaned up by the namespace controllers